Src/IronPython.Modules/_ctypes/Field.cs

```diff
@@ -112,110 +100,138 @@ public sealed class Field : PythonTypeDataSlot, ICodeFormattable {
+                        throw new InvalidOperationException("we only return int, bigint from GetValue");
                     }
+                }
+                return value;
 
-                    bits = (bits >> _bitsOffset) & validBits;
+                int ExtractBitsFromInt(int iVal) {
+                    int validBits = ((1 << _bits) - 1);
 
+                    iVal = (iVal >> _bitsOffset) & validBits;
                     if (IsSignedType) {
                         // need to sign extend if high bit is set
-                        if ((bits & (1UL << (_bits - 1))) != 0) {
-                            bits |= ulong.MaxValue ^ validBits;
+                        if ((iVal & (1 << (_bits - 1))) != 0) {
+                            iVal |= (-1) ^ validBits;
                         }
-                        value = (BigInteger)(long)bits;
-                    } else {
-                        value = (BigInteger)bits;
                     }
+
+                    return iVal;
                 }
-                return value;
             }
 
             /// <summary>
-            /// Called for fields which have been limited to a range of bits.  Sets the 
+            /// Called for fields which have been limited to a range of bits.  Sets the
             /// specified value into the bits for the field.
             /// </summary>
             private void SetBitsValue(MemoryHolder address, int baseOffset, object value) {
                 // get the value in the form of a ulong which can contain the biggest bitfield
                 ulong newBits;
-                if (value is int) {
-                    newBits = (ulong)(int)value;
-                } else if (value is BigInteger) {
-                    newBits = (ulong)(long)(BigInteger)value;
+                if (IsBoolType) {
+                    newBits = PythonOps.IsTrue(value) ? 1ul : 0ul;
                 } else {
-                    throw PythonOps.TypeErrorForTypeMismatch("int or long", value);
+                    value = PythonOps.Index(value); // since 3.8
+                    if (value is int iVal) {
+                        newBits = (ulong)iVal;
+                    } else if (value is BigInteger biVal) {
+                        if (biVal.Sign >= 0) {
+                            if (biVal <= ulong.MaxValue) {
+                                newBits = (ulong)biVal;
+                            } else {
+                                newBits = (ulong)(biVal & ulong.MaxValue);
+                            }
+                        } else {
+                            if (biVal >= long.MinValue) {
+                                newBits = (ulong)(long)biVal;
+                            } else {
+                                newBits = (ulong)(biVal & ulong.MaxValue);
+                            }
+                        }
+                    } else {
+                        throw PythonOps.TypeErrorForTypeMismatch("int", value);
```

Comment:
I think this bit is unreachable so you can throw InvalidOperationException?

Reply:
Right. It's a left-over from a pre-3.8 code version.